src/thor/route_action.cc

```diff
@@ -388,6 +388,14 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
         origin->set_date_time(origin_dt);
       }
 
+      // add waiting_secs again from the final destination's datetime, so we output the departing time
+      // at intermediate locations, not the arrival time
+      if (destination->waiting_secs() && !destination->date_time().empty()) {
+        auto dest_dt =
+            offset_date(*reader, destination->date_time(), {}, destination->waiting_secs(), {});
+        destination->set_date_time(dest_dt);
+      }
+
```

Comment:
So, now it gets a bit more complex.. We're inside the `route_two_locations` lambda of the `arrive_by` algo and go through each alternate. We already set this round's `origin.date_time` to that of the destination + elapsed secs for this path. 

Remember, on the first round we have here `origin` = input intermediate and `destination` = input destination. So there's no `waiting_secs` defined as they're not allowed on input origin/destination. After that, the current `destination` can only be an intermediate input location and potentially have `waiting_secs`. And if there were, it's `date_time` is including the `waiting_secs`, which for the reverse path means it would be serialized in the response to the `arrive_by` time at that intermediate location. However, we want the `depart_at` time there, so we add the `waiting` secs again to the current time. The only other place a location's date_time is looked at again is in the response serializer.